File: src/pdf/hittable_pdf.rs
use crate::utilities::prelude::*;
use crate::hittable::prelude::*;
use super::prelude::*;

#[derive(Clone)]
pub struct HittablePDF {
    pub hittable: Rc<dyn Hittable>,
    pub origin: P3d,
}

impl HittablePDF {
    pub fn new(hittable: Rc<dyn Hittable>, origin: P3d) -> Self {
        Self { hittable, origin }
    }
}

impl PDF for HittablePDF {
    fn value(&self, direction: Vec3) -> f64 {
        self.hittable.pdf_value(self.origin, direction)
    }

    fn generate(&self, rng: &mut SmallRng) -> Vec3 {
        self.hittable.random(self.origin, rng)
    }
}

File: lib/Pdf/HittablePdf.cpp
#include "Pdf/HittablePdf.hpp"

auto HittablePDF::value(const Vec3 &direction) const -> f64 {
    return hittable->pdf_value(origin, direction);
}
auto HittablePDF::generate() const -> Vec3 {
    return hittable->random(origin);
}
